[SPARK-4705] Handle multiple app attempts event logs, history server.

vanzin opened the pull request:
This change modifies the event logging listener to write the logs for different application
attempts to different files. The attempt ID is set by the scheduler backend, so as long
as the backend returns that ID to SparkContext, things should work. Currently, the
YARN backend does that.

The history server was also modified to model multiple attempts per application. Each
attempt has its own UI and a separate row in the listing table, so that users can look at
all the attempts separately. The UI "adapts" itself to avoid showing attempt-specific info
when all the applications being shown have a single attempt.

twinkle-g committed:
SPARK-4705: Doing cherry-pick of fix into master

Commit:
SPARK-4705 Incorporating the review comments regarding formatting, will do the rest of the changes after this

Commit:
SPARK-4705 Incorporating the review comments regarding formatting, will do the rest of the changes after this

twinkle-g committed:
SPARK-4705: 1) moved from directory structure to single file, as per the master branch. 2) Added the attempt id inside the SparkListenerApplicationStart, to make the info available independent of directory structure. 3) Changes in History Server to render the UI as per the snaphot II

Commit:
Fix my broken rebase.

Commit:
Make app attempts part of the history server model.

This change explicitly models app attempts in the history server.
An app is now a collection of app attempts, instead of the logic
to match apps to app attempts being implicit in the rendering code.

This makes the rendering code a lot simpler, since it doesn't need
to do any fancy processing of the app list to figure out what to show.

Commit:
Add a test for apps with multiple attempts.

Commit:
Attempt ID in listener event should be an option.

For backwards compatibility.

Commit:
Misc UI, test, style fixes.

Commit:
Move app name to app info, more UI fixes.

Commit:
Fix yarn history URL, app links.

vanzin commented:
This PR is an updated version of #4845. I rebased the code on top of current master, added the suggestions I made on the original PR, fixed a bunch of style nits and other issues, and added a couple of tests. Here's a screenshot:

<img src="http://i.imgur.com/EUGSRKQ.png">

SparkQA commented:
[Test build #29905 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29905/consoleFull) for   PR 5432 at commit [`657ec18`](https://github.com/apache/spark/commit/657ec18b5affa0a3f77f51a3e3e05ddbbc21c85f).

SparkQA commented:
[Test build #29905 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29905/consoleFull) for   PR 5432 at commit [`657ec18`](https://github.com/apache/spark/commit/657ec18b5affa0a3f77f51a3e3e05ddbbc21c85f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29905/
Test FAILed.

Commit:
Argh scalastyle.

SparkQA commented:
[Test build #29907 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29907/consoleFull) for   PR 5432 at commit [`3a14503`](https://github.com/apache/spark/commit/3a145030c7a2f729ce9665fdd454a83b21750c16).

vanzin commented:
BTW the zebra-striping in the UI looks a little broken right now, I'll take a look at that.

SparkQA commented:
**[Test build #29907 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29907/consoleFull)**     for PR 5432 at commit [`3a14503`](https://github.com/apache/spark/commit/3a145030c7a2f729ce9665fdd454a83b21750c16)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29907/
Test FAILed.

vanzin commented:
Hmm, didn't find a test failure in the output.

vanzin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #29917 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29917/consoleFull) for   PR 5432 at commit [`3a14503`](https://github.com/apache/spark/commit/3a145030c7a2f729ce9665fdd454a83b21750c16).

SparkQA commented:
[Test build #29917 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29917/consoleFull) for   PR 5432 at commit [`3a14503`](https://github.com/apache/spark/commit/3a145030c7a2f729ce9665fdd454a83b21750c16).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29917/
Test FAILed.

Commit:
Fix HistoryServer test.

SparkQA commented:
[Test build #29949 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29949/consoleFull) for   PR 5432 at commit [`9092af5`](https://github.com/apache/spark/commit/9092af517c07504a275f7d6ae2187fe97f8c83cd).

SparkQA commented:
[Test build #29949 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29949/consoleFull) for   PR 5432 at commit [`9092af5`](https://github.com/apache/spark/commit/9092af517c07504a275f7d6ae2187fe97f8c83cd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29949/
Test FAILed.

vanzin commented:
Funny. All YARN tests (not just in this PR) are failing with this:

    /bin/bash: /bin/java: No such file or directory

Wonder what changed in the environment since they were working before? (And why is github's user name search so useless it cannot autocomplete Shane's user name?)

shaneknapp commented:
so i just grepped through the code and found stuff like this:

yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:        YarnSparkHadoopUtil.expandEnvironment(Environment.JAVA_HOME) + "/bin/java", "-server"
yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala:      YarnSparkHadoopUtil.expandEnvironment(Environment.JAVA_HOME) + "/bin/java",

i've never explicitly set JAVA_HOME in jenkins' slave user space before, but that's obviously why it's failing.  that's pretty bad code imo.

solutions:
* explicitly set JAVA_HOME in each slave's config (bad, as it ties that slave to whatever is on system java)
* if JAVA_HOME isn't set, use whatever java is in the path (good)
* explicitly define which java version to test against in the jenkins build's config

shaneknapp commented:
btw i prefer #2.

vanzin commented:
I think `JAVA_HOME` is something that YARN exposes to all containers, so even if you don't set it for your application, that code should still work. I think we problem here is a little different - we should just make sure the tests have the same env as you'd find in an usual YARN installation.

Anyway, I'm trying something out in #5441.

shaneknapp commented:
cool.  on our systems, at least, the system java we use is /usr/bin/java, which points (through /etc/alternatives), to /usr/java/latest (which itself is a link to /usr/java/jdk1.7.0_71/).

shaneknapp commented:
oh, i just had a thought:  i installed a couple of different versions of java through jenkins, and right now the tests are set in the config to use 'Default', which is system level java.  i bet this is why JAVA_HOME isn't being set and why the tests are failing.  @JoshRosen is going to set JAVA_HOME for us to get the builds green and then we can look a little deeper in to the problem.

andrewor14 commented:
Is it always safe to rely on `java.home` pointing to the right directory? IIUC this is independently of whether we use Maven or SBT. Then perhaps the correct way of fixing this is doing something like what `AbstractCommandBuilder` does, where if `JAVA_HOME` is not set it defaults to using `java.home`:https://github.com/apache/spark/blob/470d7453a56c56a41b2851551fe1830065f88b2c/launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java#L93

vanzin commented:
Note that the YARN code is not resolving `JAVA_HOME` locally, it's adding a reference to `$JAVA_HOME` to the command that will be executed by YARN. That will be resolved on the node where the command is run. The NM generally sets `JAVA_HOME` for child processes.

andrewor14 commented:
Ah yes, scratch that.

vanzin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #29996 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29996/consoleFull) for   PR 5432 at commit [`9092af5`](https://github.com/apache/spark/commit/9092af517c07504a275f7d6ae2187fe97f8c83cd).

SparkQA commented:
[Test build #29996 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29996/consoleFull) for   PR 5432 at commit [`9092af5`](https://github.com/apache/spark/commit/9092af517c07504a275f7d6ae2187fe97f8c83cd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29996/
Test PASSed.

Commit:
Disable striping for app id / name when multiple attempts exist.

SparkQA commented:
[Test build #30043 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30043/consoleFull) for   PR 5432 at commit [`07446c6`](https://github.com/apache/spark/commit/07446c688f94ccdb98645b2e75960ab745caa339).

SparkQA commented:
[Test build #30043 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30043/consoleFull) for   PR 5432 at commit [`07446c6`](https://github.com/apache/spark/commit/07446c688f94ccdb98645b2e75960ab745caa339).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30043/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-4705

Conflicts:
	yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

SparkQA commented:
[Test build #30209 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30209/consoleFull) for   PR 5432 at commit [`86de638`](https://github.com/apache/spark/commit/86de638c3579fb2d5ef07548003d80e835fb2c18).

SparkQA commented:
[Test build #30209 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30209/consoleFull) for   PR 5432 at commit [`86de638`](https://github.com/apache/spark/commit/86de638c3579fb2d5ef07548003d80e835fb2c18).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30209/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-4705

Conflicts:
	core/src/main/scala/org/apache/spark/SparkContext.scala

SparkQA commented:
[Test build #30496 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30496/consoleFull) for   PR 5432 at commit [`9092d39`](https://github.com/apache/spark/commit/9092d39bf660f096f1a74f76d42a9760b415f3a1).

vanzin commented:
/cc @andrewor14 @harishreedharan it would be great to have some eyes on this, thanks!

SparkQA commented:
[Test build #30496 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30496/consoleFull) for   PR 5432 at commit [`9092d39`](https://github.com/apache/spark/commit/9092d39bf660f096f1a74f76d42a9760b415f3a1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30496/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-4705

Conflicts:
	yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

SparkQA commented:
[Test build #30599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30599/consoleFull) for   PR 5432 at commit [`c14ec19`](https://github.com/apache/spark/commit/c14ec19c9e54a2ef1920a6d0f465d7913f5b8299).

SparkQA commented:
[Test build #30599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30599/consoleFull) for   PR 5432 at commit [`c14ec19`](https://github.com/apache/spark/commit/c14ec19c9e54a2ef1920a6d0f465d7913f5b8299).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`
  * `public class MapConfigProvider extends ConfigProvider `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30599/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/scheduler/SchedulerBackend.scala:
```diff
@@ -41,4 +41,11 @@ private[spark] trait SchedulerBackend {
+   *
+   * @return An application attempt id
+   */
+  def applicationAttemptId(): String = ""
```
I think it would be cleaner if you used an `Option[String]` with a default of `None`.  As I was reading the rest of the code I was wondering if every framework would be forced to provide some dummy attempt id, or if would be null, etc.  `Option` would make that more explicit.  Plus it seems you use an option later in the listener event for json in any case.

vanzin replied:
Yeah, I went back and forth on this bug forgot to revisit the code. I agree that `Option` would be better here, let me see how big a change that would be.

squito commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -194,7 +194,8 @@ private[spark] object JsonProtocol {
     ("Timestamp" -> applicationStart.time) ~
-    ("User" -> applicationStart.sparkUser)
+    ("User" -> applicationStart.sparkUser) ~
+    ("appAttemptId" -> applicationStart.appAttemptId.map(JString(_)).getOrElse(JNothing))
```
You should add a test case here: 
https://github.com/apache/spark/blob/fbe7106d75c6a1624d10793fba6759703bc5c6e6/core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala#L275

Commit:
Merge branch 'master' into SPARK-4705

Commit:
Use Option[String] for attempt id.

Commit:
Update JsonProtocol test, make property name consistent.

SparkQA commented:
[Test build #30863 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30863/consoleFull) for   PR 5432 at commit [`d5a9c37`](https://github.com/apache/spark/commit/d5a9c37a00f3b0b5aa66c5f92c325fdf0ac05bf0).

SparkQA commented:
[Test build #30863 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30863/consoleFull) for   PR 5432 at commit [`d5a9c37`](https://github.com/apache/spark/commit/d5a9c37a00f3b0b5aa66c5f92c325fdf0ac05bf0).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30863/
Test FAILed.

Commit:
Scalastyle...

SparkQA commented:
[Test build #30870 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30870/consoleFull) for   PR 5432 at commit [`9d59d92`](https://github.com/apache/spark/commit/9d59d92deff81863c8de58f8909bd0cf929b60c3).

SparkQA commented:
[Test build #30870 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30870/consoleFull) for   PR 5432 at commit [`9d59d92`](https://github.com/apache/spark/commit/9d59d92deff81863c8de58f8909bd0cf929b60c3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30870/
Test FAILed.

Commit:
Missed a reference to the old property name.

squito commented:
just chatted w/ Marcelo about this offline for a bit, one issue here is how apps get displayed when there are some complete and some running attempts -- which page do they go on, are they all grouped together, and what is the sort order within attempts?  Marcelo is going to follow up on that.

Other than that one issue, this is looking good, I'll take another pass after those changes.

SparkQA commented:
[Test build #30876 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30876/consoleFull) for   PR 5432 at commit [`2ad77e7`](https://github.com/apache/spark/commit/2ad77e7f93aec5cc0f9d1b0136352fe22bab5683).

Commit:
Improve sorting of app attempts.

Make sure running attempts show up first. Also, when an app has
running attempts, don't show it in the complete app list.

SparkQA commented:
[Test build #30880 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30880/consoleFull) for   PR 5432 at commit [`1aa309d`](https://github.com/apache/spark/commit/1aa309d4a3b29fc106629f035ce24c604853bf1c).

SparkQA commented:
[Test build #30876 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30876/consoleFull) for   PR 5432 at commit [`2ad77e7`](https://github.com/apache/spark/commit/2ad77e7f93aec5cc0f9d1b0136352fe22bab5683).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30876/
Test PASSed.

SparkQA commented:
[Test build #30880 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30880/consoleFull) for   PR 5432 at commit [`1aa309d`](https://github.com/apache/spark/commit/1aa309d4a3b29fc106629f035ce24c604853bf1c).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30880/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-4705

Conflicts:
	core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

vanzin commented:
Hmmm... I'll have to fix the log cleaner. Shouldn't be terribly hard, and I'll also add tests for it on top, but feel free to look at the rest of the code in the meantime.

Commit:
Fix log cleaner, add test.

squito commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -232,76 +239,105 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
+        } else if (toClean.size < app.attempts.size) {
+          appsToRetain += (app.id ->
+            new FsApplicationHistoryInfo(app.id, app.name,
+              app.attempts.filter(!shouldClean(_)).toList))
```
nit: instead of filtering twice, you can do
```
val  (toClean, toKeep) = app.attempts.partition(shouldClean)
```

vanzin replied:
I swear I read the whole API for `Seq` and missed that one...

squito commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryPage.scala:
```diff
@@ -34,18 +37,28 @@ private[history] class HistoryPage(parent: HistoryServer) extends WebUIPage("")
+    val allAppsSize = allApps.size
+
+    val actualFirst = if (requestedFirst < allAppsSize) requestedFirst else 0
+    val appsToShow = allApps.slice(actualFirst, Math.min(actualFirst + pageSize, allAppsSize))
```
I realize you didn't add this, but the `Math.min` isn't necessary:

```
scala> (1 to 10).slice(5,20)
res0: scala.collection.immutable.IndexedSeq[Int] = Vector(6, 7, 8, 9, 10)
```

squito commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryPage.scala:
```diff
@@ -125,30 +138,85 @@ private[history] class HistoryPage(parent: HistoryServer) extends WebUIPage("")
+        UIUtils.formatDuration(attempt.endTime - attempt.startTime)
+        } else {
+          "-"
+        }
```
nit: fix indentation

squito commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -159,52 +167,52 @@ class FsHistoryProviderSuite extends FunSuite with BeforeAndAfter with Matchers
     writeFile(logFile1, true, None,
-      SparkListenerApplicationStart("app1", Some("app1"), 1L, "test"),
+      SparkListenerApplicationStart("app1", Some("app1"), 1L, "test", None),
       SparkListenerApplicationEnd(2L))
```
also not your change, but -- doesn't `inProgress = true` imply no `SparkListenerApplicationEnd` event?

vanzin replied:
Not necessarily. `inProgress = true` just affects the log file name.

squito commented on core/src/main/scala/org/apache/spark/deploy/history/ApplicationHistoryProvider.scala:
```diff
@@ -41,9 +45,11 @@ private[history] abstract class ApplicationHistoryProvider {
    * @return The application's UI, or None if application is not found.
    */
-  def getAppUI(appId: String): Option[SparkUI]
+  def getAppUI(appId: String, attemptId: String): Option[SparkUI]
```
any reason not to use an `Option[String]` for `attemptId` here (and everywhere else) as well?

squito commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -138,31 +142,34 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
-          // Do not call ui.bind() to avoid creating a new server for each application
-        }
+      applications.get(appId).flatMap { appInfo =>
+        val attempts = appInfo.attempts.filter(_.attemptId == attemptId)
```
the doc for `getAppUI` says to use an empty string for apps with a single attempt -- but that isn't exactly what is reflected here.  Some yarn apps will be successful on the first attempt, but with this implementation, you still need to pass in the actual attempt id.  One way or the other, the doc & this should be resolved.

vanzin replied:
The interface doc is slightly misleading, but all event logs from YARN will have an attempt ID after this change, even for a single attempt.

Commit:
Review feedback.

SparkQA commented:
[Test build #31146 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31146/consoleFull) for   PR 5432 at commit [`bc885b7`](https://github.com/apache/spark/commit/bc885b7295ade58f1e0c3159d4e5a2cd7658ef3b).

SparkQA commented:
[Test build #31146 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31146/consoleFull) for   PR 5432 at commit [`bc885b7`](https://github.com/apache/spark/commit/bc885b7295ade58f1e0c3159d4e5a2cd7658ef3b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31146/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-4705

Conflicts:
	core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31166 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31166/consoleFull) for   PR 5432 at commit [`f66dcc5`](https://github.com/apache/spark/commit/f66dcc5b0625910d522de184b25800ecdf456366).

squito commented:
lgtm

I'll wait a day for any other feedback.

andrewor14 commented:
@vanzin thanks for the fix. I'll have a quick look at this tonight.

SparkQA commented:
[Test build #31166 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31166/consoleFull) for   PR 5432 at commit [`f66dcc5`](https://github.com/apache/spark/commit/f66dcc5b0625910d522de184b25800ecdf456366).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31166/
Test PASSed.

squito commented:
@andrewor14 did you have any comments on this?  otherwise I am ready to merge

andrewor14 commented:
Sorry I will look at this right now.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -138,31 +142,34 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
-        }
+      applications.get(appId).flatMap { appInfo =>
+        val attempts = appInfo.attempts.filter(_.attemptId == attemptId)
+        attempts.headOption.map { attempt =>
```
minor: this can just be `find` I think

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryPage.scala:
```diff
@@ -22,6 +22,9 @@ import javax.servlet.http.HttpServletRequest
 import org.apache.spark.ui.{WebUIPage, UIUtils}
+import scala.collection.immutable.ListMap
+import scala.collection.mutable.HashMap
+import scala.collection.mutable.ArrayBuffer
```
import order

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala:
```diff
@@ -76,18 +80,23 @@ class HistoryServer(
-      val appId = parts(1)
+      val appKey =
+        if (parts.length == 3) {
+          s"${parts(1)}/${parts(2)}"
```
can you add a comment on what these parts represent? Maybe add an example in the comment?

andrewor14 replied:
oh I see. IIUC this corresponds to `getAttemptURI` below

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -261,11 +267,20 @@ private[spark] object EventLoggingListener extends Logging {
       appId: String,
+      appAttemptId: Option[String],
       compressionCodecName: Option[String] = None): String = {
     val sanitizedAppId = appId.replaceAll("[ :/]", "-").replaceAll("[.${}'\"]", "_").toLowerCase
```
can we call `sanitize` on this too?

andrewor14 replied:
actually I don't think this variable is used

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/SchedulerBackend.scala:
```diff
@@ -41,4 +41,11 @@ private[spark] trait SchedulerBackend {
 
+  /**
+   * Get an application ID associated with the job.
+   *
```
when is this defined vs None? Is it as simple as "if the cluster manager provides it then it's defined, otherwise none"?

andrewor14 replied:
might be worth a comment even though that is the case the developer doesn't need to guess.

andrewor14 replied:
actually, does it make sense for applications running in client mode to have an attempt ID?

andrewor14 commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -194,7 +194,8 @@ private[spark] object JsonProtocol {
     ("Timestamp" -> applicationStart.time) ~
-    ("User" -> applicationStart.sparkUser)
+    ("User" -> applicationStart.sparkUser) ~
+    ("App Attempt ID" -> applicationStart.appAttemptId.map(JString(_)).getOrElse(JNothing))
```
super minor but I would move this right under `App ID` since they're logically related.

andrewor14 commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -96,33 +99,39 @@ class FsHistoryProviderSuite extends FunSuite with BeforeAndAfter with Matchers
+      list.count(_.attempts.head.completed) should be (3)
+
+      def makeAppInfo(id: String, name: String, start: Long, end: Long, lastMod: Long,
+        user: String, completed: Boolean): ApplicationHistoryInfo = {
```
can you format this properly

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -89,6 +89,10 @@ private[spark] class ApplicationMaster(
+
+        // Propagate the attempt if, so that in case of event logging,
+        // different attempt's logs gets created in different directory
+        System.setProperty("spark.yarn.app.attemptid", appAttemptId.getAttemptId().toString())
```
should this be `attemptId` camel case?

andrewor14 commented on yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClusterSchedulerBackend.scala:
```diff
@@ -47,4 +47,12 @@ private[spark] class YarnClusterSchedulerBackend(
+  override def applicationAttemptId(): Option[String] =
+    // In YARN Cluster mode, spark.yarn.app.attemptid is expect to be set
+    // before user application is launched.
+    // So, if spark.yarn.app.id is not set, it is something wrong.
```
is this supposed to be `spark.yarn.app.attemptId` instead of just the `app.id`? Maybe a simpler way to put this is "The attempt ID is always set for YARN cluster applications"

andrewor14 commented:
Great to see this fixed @vanzin. My comments are mostly minor and unfortunately I don't have the time to do a closer review. How much more work do you imagine fixing this additionally for standalone mode would be? If it's not that much we should also fix that for 1.4 in separate patch.

vanzin commented:
> How much more work do you imagine fixing this additionally for standalone mode would be?

I have no idea, I'm mostly unfamiliar with standalone cluster mode. Feel free to file a separate bug for it.

Commit:
Review feedback.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31464 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31464/consoleFull) for   PR 5432 at commit [`7e289fa`](https://github.com/apache/spark/commit/7e289fa1297aa21f6c8764e4a237eb3a674675e0).

SparkQA commented:
[Test build #31464 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31464/consoleFull) for   PR 5432 at commit [`7e289fa`](https://github.com/apache/spark/commit/7e289fa1297aa21f6c8764e4a237eb3a674675e0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31464/
Test FAILed.

andrewor14 commented:
retest this please

andrewor14 commented:
Latest changes LGTM based on my reviews. @squito feel free to merge it.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31480 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31480/consoleFull) for   PR 5432 at commit [`7e289fa`](https://github.com/apache/spark/commit/7e289fa1297aa21f6c8764e4a237eb3a674675e0).

SparkQA commented:
[Test build #31480 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31480/consoleFull) for   PR 5432 at commit [`7e289fa`](https://github.com/apache/spark/commit/7e289fa1297aa21f6c8764e4a237eb3a674675e0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SparkListenerApplicationStart(appName: String, appId: Option[String],`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31480/
Test PASSed.
